src/model/immutable/SelectionState.ts

```diff
@@ -87,6 +87,46 @@ export function hasEdgeWithin(
   return start <= offsetToCheck && end >= offsetToCheck;
 }
 
+/**
+ * Returns whether the selection starts (if forward) or ends (if backward)
+ * on the trailing edge of a leaf node, and if there are additional leaves
+ * in the block after it.
+ */
+export function isOnlyOnTrailingEdgeAndIsNotLastInBlock(
+  sel: SelectionState,
+  blockKey: string,
+  end: number,
+  blockLength: number,
+): boolean {
+  const {isBackward, anchorOffset, focusOffset, focusKey, anchorKey} = sel;
+  return (
+    (isBackward && focusKey === blockKey && focusOffset === end && end < blockLength) ||
+    (!isBackward && anchorKey === blockKey && anchorOffset === end && end < blockLength)
+  );
+}
+
+/**
+ * Returns whether the selection ends (if forward) or starts (if backward)
+ * on the leading edge of a leaf node, and if there are additional leaves
+ * in the block _and in the selection_ before it.
```

Comment:
It's also not clear to me why here we require that the preceding leaves are in the selection, but in the function above we don't require the succeeding ones are in selection?

Reply:
Selection is processed in leaf-order. The main problem these functions are trying to solve is that there are 2 ways to represent selection if it falls on the edge of a node: the end of node `n`, or the beginning of node `n+1`. These functions standardize the approach by always biasing for the latter. But if there _is_ no node `n+1`, we should bias toward the former.

I'll make this more clear in the documentation, possibly with a diagram.